### PR TITLE
RBAC: extend `prompt` resource_type to after-request handlers

### PR DESCRIPTION
**Pull request by PattaraS**
### Related Issues/PRs

Relates to #23248

### What changes are proposed in this pull request?

Follow-up to #23248 (which promoted `prompt` to a first-class RBAC `resource_type`). The before-request dispatchers were already prompt-aware, but four after-request handlers still operated in the `registered_model` namespace exclusively — flagged in NaiLaOpus's review after merge:

| Handler | Gap before this PR |
|---|---|
| `filter_search_registered_models` (`__init__.py:2845`) | `_role_based_read_predicate(username, "registered_model")` only — a user holding `(prompt, foo, READ)` had prompt `foo` silently filtered out of search results. |
| `filter_search_model_versions` (`__init__.py:2893`) | Same gap for prompt-backed model versions. |
| `rename_registered_model_permission` (`__init__.py:2909`) | `rename_grants_for_resource("registered_model", ...)` only — renaming a prompt orphaned `(prompt, old_name, ...)` grants. |
| `delete_can_manage_registered_model_permission` (`__init__.py:2432`) | `delete_grants_for_resource("registered_model", ...)` only — deleting a prompt leaked `(prompt, name, ...)` rows permanently. |

**Search filters** now classify each row by its persisted `mlflow.prompt.is_prompt` tag (proto-side via new `_proto_is_prompt`; ORM-side via `RegisteredModel._is_prompt()`) and dispatch to the matching predicate. Cross-namespace grants do not leak through in either direction.

**Rename and delete** sweep both `registered_model` and `prompt` namespaces unconditionally. Names are unique within the model registry, so exactly one of the two sweeps applies and the other is a no-op — avoids needing to classify at after-request time, when the entity is already gone (delete) or renamed (rename).

### How is this PR tested?

- [x] New unit/integration tests
- [ ] Manual tests

Five tests in `tests/server/auth/test_auth_workspace.py`:
- `test_filter_search_registered_models_uses_prompt_grant_for_prompt_rows` — prompt READ grant satisfies prompt row.
- `test_filter_search_registered_models_does_not_satisfy_prompt_with_rm_grant` — `registered_model` grant does NOT leak to prompt namespace (regression guard).
- `test_filter_search_model_versions_uses_prompt_grant_for_prompt_versions` — `ModelVersion` mirror.
- `test_rename_registered_model_permission_sweeps_prompt_namespace` — rename propagates to `(prompt, ...)` grants.
- `test_delete_can_manage_registered_model_permission_sweeps_prompt_namespace` — delete sweeps `(prompt, ...)` grants.

Full `tests/server/auth/test_auth_workspace.py` is green (165/165). Ruff + Clint clean.

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

RBAC: prompt resource-type now applies consistently across all after-request handlers. Searches return prompt rows for users with `prompt` grants, and prompt renames / deletes propagate to RBAC grants.

#### What component(s), interfaces, languages, and integrations does this PR affect?

- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

---
<a href="https://app.blacksmith.sh/mlflow/codesmith/mlflow/pr/23426"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

### CODE REVIEW

**Comment by PattaraS** [comment]
/review-v2

---

🚀 [NaiLaOpus running...](https://github.com/mlflow/mlflow/actions/runs/26029281720)

File: tests/server/auth/test_auth_workspace.py [hunk 1]
```diff
@@ -1253,6 +1255,195 @@ def get_registered_model(self, name):
...
+
+def _build_search_model_versions_response(entries):
+    """Build a ``SearchModelVersions.Response()`` from ``(name, is_prompt)`` pairs."""
+    resp = SearchModelVersions.Response()
+    for name, is_prompt in entries:
+        mv = resp.model_versions.add()
+        mv.name = name
+        if is_prompt:
+            mv.tags.add(key=IS_PROMPT_TAG_KEY, value="true")
+    return resp
```
[the remark this answers is not in the document]
Comment (PattaraS) [reply tests/server/auth/test_auth_workspace.py]: Done in `6a9a92d38` — removed both `_build_search_*_response` helpers and the unused `SearchModelVersions`/`SearchRegisteredModels` proto imports. The tests use raw JSON payloads (mimicking what `parse_dict` would see) so the helpers ended up dead code.

🤖 *Claude response*

File: mlflow/server/auth/__init__.py [hunk 2]
```diff
@@ -2424,12 +2433,19 @@ def set_can_manage_registered_model_permission(resp: Response):
...
+
+    The model registry's primary key is the entity name (unlike experiments
+    which use a UUID), so a future entity with the same name would otherwise
+    inherit stale grants. ``DeleteRegisteredModel`` is shared between
+    registered models and prompts on the REST surface; the entity is already
+    gone by the time this after-request hook runs, so we cannot classify it
+    now. Names are unique within the registry, so exactly one of the two
+    sweeps applies and the other is a no-op.
     """
     name = request.get_json(force=True, silent=True)["name"]
```
[the remark this answers is not in the document]
Comment (PattaraS) [reply mlflow/server/auth/__init__.py]: Fixed in `6a9a92d38` — added `or {}` guard and an explicit `name` check that raises `INVALID_PARAMETER_VALUE` instead of bubbling a `TypeError` to 500. New test `test_delete_can_manage_registered_model_permission_rejects_missing_name` pins the 400 path.

🤖 *Claude response*

File: mlflow/server/auth/__init__.py [hunk 3]
```diff
@@ -2890,28 +2919,34 @@ def filter_search_model_versions(resp: Response):
...
-    A model registry can be assigned to multiple users with different permissions.
+    Propagate a registered-model rename to RBAC grants.
 
-    Changing the model registry name must be propagated to all users.
+    ``RenameRegisteredModel`` is shared between registered models and prompts;
+    sweep both namespaces so a prompt rename doesn't orphan its
+    ``(prompt, old_name, ...)`` grants. Names are unique within the registry,
+    so exactly one of the two renames applies and the other is a no-op.
     """
     data = request.get_json(force=True, silent=True)
```
[the remark this answers is not in the document]
Comment (PattaraS) [reply mlflow/server/auth/__init__.py]: Fixed in `6a9a92d38` — added `or {}` plus explicit `name` / `new_name` validation that raises `INVALID_PARAMETER_VALUE` when either is absent. Without this, `None` would propagate to `rename_grants_for_resource(resource_pattern=None)`, which the agreement above pointed out could silently corrupt grants. New test `test_rename_registered_model_permission_rejects_missing_fields` pins the rejection.

🤖 *Claude response*

File: mlflow/server/auth/__init__.py [hunk 4]
```diff
@@ -2869,7 +2894,11 @@ def filter_search_registered_models(resp: Response):
             response_message.next_page_token = ""
             break
 
-        refetched_readable_proto = [rm.to_proto() for rm in refetched if can_read(rm.name)]
+        refetched_readable_proto = [
+            rm.to_proto()
+            for rm in refetched
+            if (can_read_prompt if rm._is_prompt() else can_read_rm)(rm.name)
```
[the remark this answers is not in the document]
Comment (PattaraS) [reply mlflow/server/auth/__init__.py]: Fixed in `6a9a92d38` — unified by extracting `_entity_is_prompt(entity)` that dispatches via `_is_prompt()` for ORM entities (necessary because `RegisteredModel.tags` strips the `mlflow.prompt.is_prompt` key from its public dict) and falls back to a `.tags` scan for protos / `ModelVersion` ORM (whose `.tags` is the raw dict). Both call sites in `filter_search_registered_models` (initial + refetch loop) and `filter_search_model_versions` now share one classifier.

🤖 *Claude response*

File: tests/server/auth/test_auth_workspace.py [hunk 5]
```diff
@@ -1253,6 +1255,195 @@ def get_registered_model(self, name):
...
+
+
+def test_filter_search_registered_models_uses_prompt_grant_for_prompt_rows(
```
[the remark this answers is not in the document]
Comment (PattaraS) [reply tests/server/auth/test_auth_workspace.py:1256]: Done in `6a9a92d38` — added `test_filter_search_registered_models_classifies_refetched_rows`. It seeds an empty initial response with a valid `next_page_token`, monkeypatches `_get_model_registry_store` to return a `PagedList` of ORM `RegisteredModel` entities (one flagged with `mlflow.prompt.is_prompt=true`, one not), and asserts the prompt row passes the filter via the prompt grant. This pins the ORM-side classification path that the original tests missed.

🤖 *Claude response*

**Comment by PattaraS** [comment]
Re: [NaiLaOpus comment on `__init__.py:2431`](https://github.com/mlflow/mlflow/pull/23426#issuecomment-4477011133):

Fixed in `6a9a92d38` — `set_can_manage_registered_model_permission` now inspects `response_message.registered_model.tags` via the new `_entity_is_prompt` helper and grants MANAGE in the matching namespace. Two new tests pin both paths:

- `test_set_can_manage_registered_model_permission_grants_prompt_for_prompt_entity` — a prompt creation lands a `(prompt, name, MANAGE)` grant and no `(registered_model, name, MANAGE)` grant.
- `test_set_can_manage_registered_model_permission_grants_registered_model_for_plain_entity` — regression guard that the new dispatch path doesn't flip the namespace for normal registered models.

Thanks for catching this — it would have left every prompt creator immediately locked out by the prompt-side validators.

🤖 *Claude response*

**Comment by PattaraS** [comment]
/review-v2 --no-cache

---

🚀 [NaiLaOpus running...](https://github.com/mlflow/mlflow/actions/runs/26042840789)

**Comment by PattaraS** [comment]
/review

---

✅ [Review](https://github.com/mlflow/mlflow/actions/runs/26043087608?pr=23426) completed (claude-opus-4-6, 140.4s, 16 turns, 4042 tokens, $0.55)
<details><summary>Result</summary>

```json
{
  "type": "result",
  "subtype": "success",
  "is_error": false,
  "api_error_status": null,
  "duration_ms": 140446,
  "duration_api_ms": 134922,
  "num_turns": 16,
  "result": "Review payload validated. Here's my assessment:\n\n**APPROVE** — This PR correctly extends RBAC prompt support to all four after-request handlers that were previously operating in the `registered_model` namespace only.\n\nKey observations:\n- **`_entity_is_prompt`** is well-designed: it dispatches via `_is_prompt()` for ORM `RegisteredModel` (whose `.tags` hides the prompt flag) and falls back to tag scanning for protos and ORM `ModelVersion` (whose `.tags` is a plain dict). All three entity shapes are handled correctly.\n- **Dual-namespace sweeps** for rename/delete are the right call — the entity is gone or renamed by the time these after-request hooks fire, so classification isn't possible. The no-op sweep is harmless.\n- **Input validation** (`or {}` guards + explicit `MlflowException` raises) properly addresses the `None`/`TypeError` bugs flagged in prior review.\n- **Test coverage** is thorough: 10 new tests covering prompt grant satisfaction, cross-namespace isolation, the ORM refetch path, create-time classification, and input validation edge cases.\n\nOne minor NIT filed: `filter_search_model_versions` uses a slightly different dispatch pattern than `filter_search_registered_models` — could be unified for consistency.",
  "stop_reason": "end_turn",
  "session_id": "a6696f85-d1ba-4374-9721-92f016330715",
  "total_cost_usd": 0.5513717499999999,
  "usage": {
    "input_tokens": 12,
    "cache_creation_input_tokens": 36803,
    "cache_read_input_tokens": 320639,
    "output_tokens": 4030,
    "server_tool_use": {
      "web_search_requests": 0,
      "web_fetch_requests": 0
    },
    "service_tier": "standard",
    "cache_creation": {
      "ephemeral_1h_input_tokens": 0,
      "ephemeral_5m_input_tokens": 36803
    },
    "inference_geo": "",
    "iterations": [
      {
        "input_tokens": 1,
        "output_tokens": 304,
        "cache_read_input_tokens": 49758,
        "cache_creation_input_tokens": 111,
        "cache_creation": {
          "ephemeral_5m_input_tokens": 111,
          "ephemeral_1h_input_tokens": 0
        },
        "type": "message"
      }
    ],
    "speed": "standard"
  },
  "modelUsage": {
    "claude-haiku-4-5-20251001": {
      "inputTokens": 398,
      "outputTokens": 1764,
      "cacheReadInputTokens": 150680,
      "cacheCreationInputTokens": 28750,
      "webSearchRequests": 0,
      "costUSD": 0.0602235,
      "contextWindow": 200000,
      "maxOutputTokens": 32000
    },
    "claude-opus-4-6": {
      "inputTokens": 12,
      "outputTokens": 4030,
      "cacheReadInputTokens": 320639,
      "cacheCreationInputTokens": 36803,
      "webSearchRequests": 0,
      "costUSD": 0.49114825,
      "contextWindow": 200000,
      "maxOutputTokens": 64000
    }
  },
  "permission_denials": [],
  "terminal_reason": "completed",
  "fast_mode_state": "off",
  "uuid": "32bd1abd-8ca9-4f51-907b-6d11fa6ff9b7"
}
```

</details>

File: mlflow/server/auth/__init__.py [hunk 6]
```diff
@@ -3093,28 +3148,42 @@ def filter_search_model_versions(resp: Response):
...
+    # Prompt versions and model versions share the same REST surface; classify
+    # each row by its ``mlflow.prompt.is_prompt`` tag so a prompt-version
+    # carrying a ``(prompt, name, READ)`` grant isn't dropped on the floor.
+    can_read_rm = _role_based_read_predicate(username, "registered_model")
+    can_read_prompt = _role_based_read_predicate(username, "prompt")
     # filter out model versions whose parent model is unreadable
     for mv in list(response_message.model_versions):
-        if not can_read(mv.name):
+        check = can_read_prompt if _entity_is_prompt(mv) else can_read_rm
+        if not check(mv.name):
```
[the remark this answers is not in the document]
Comment (PattaraS) [reply mlflow/server/auth/__init__.py]: Done in `1a425a77d` — `filter_search_model_versions` now uses the same `can_read(entity)` closure pattern as `filter_search_registered_models`. The two filters are structurally identical now.

🤖 *Claude response*

File: mlflow/server/auth/__init__.py [hunk 7]
```diff
@@ -2971,10 +3014,19 @@ def filter_search_registered_models(resp: Response):
...
-    can_read = _role_based_read_predicate(username, "registered_model")
+    # The registered-model REST surface is shared with prompts; classify each
+    # row by its ``mlflow.prompt.is_prompt`` tag and check the correct grant
+    # namespace. Without this, a user holding only a ``(prompt, foo, READ)``
+    # grant would have prompt ``foo`` silently filtered out of the response.
+    can_read_rm = _role_based_read_predicate(username, "registered_model")
+    can_read_prompt = _role_based_read_predicate(username, "prompt")
+
+    def can_read(entity) -> bool:
+        return (can_read_prompt if _entity_is_prompt(entity) else can_read_rm)(entity.name)
```
Comment (B-Step62) [inline mlflow/server/auth/__init__.py]: nit: I guess these doesn't need to be inline?

File: mlflow/server/auth/__init__.py [hunk 8]
```diff
@@ -1484,6 +1485,24 @@ def validate_can_get_user_permission() -> bool:
...
+    Unifies the two response-filtering paths in ``filter_search_*`` — initial
+    response rows arrive as protos (via ``parse_dict``), refetched rows arrive
+    as ORM entities (``PagedList[RegisteredModel]``). Dispatch order matters:
+    ``RegisteredModel.tags`` strips the ``mlflow.prompt.is_prompt`` tag from
+    its public dict so users don't see it, so we must call ``_is_prompt()``
+    on that entity rather than reading ``.tags`` directly.
+    """
+    if hasattr(entity, "_is_prompt"):
+        return entity._is_prompt()
+    tags = entity.tags
```
Comment (B-Step62) [inline mlflow/server/auth/__init__.py]: nit: I guess this tag branch is for PromptVersion? Shall we add similar `_is_prompt()` method to the ModelVersion entity?

**Review by B-Step62 — APPROVED** [review]
LGTM with two nits!

File: mlflow/server/auth/__init__.py (hunk 7 above)
Comment (PattaraS) [reply mlflow/server/auth/__init__.py]: Good call — fixed in `89b68077d`. Both filters had the same inline closure now that they're structurally identical; extracted to a module-level helper `_rm_or_prompt_read_predicate(username)` and the two filters each just call `can_read = _rm_or_prompt_read_predicate(username)` now.

🤖 *Claude response*

File: mlflow/server/auth/__init__.py (hunk 8 above)
Comment (PattaraS) [reply mlflow/server/auth/__init__.py]: Yep — done in `89b68077d`. Added `ModelVersion._is_prompt()` mirroring the existing `RegisteredModel._is_prompt()`, and dropped the `isinstance(tags, dict)` branch from `_entity_is_prompt` (it became dead code since both ORM entities now carry the method). The helper is now a clean 2-branch dispatch: ORM method, else proto tag-scan.

🤖 *Claude response*
